### PR TITLE
rfc: resampling with no pixel shift

### PR DESCRIPTION
**Pull request by kamil-andrzejewski**
Default linear algorithm in TF 1.x behaves differently than in version 2.x. OneDNN currently supports a version compatible with TF 2.x. The expectation is to have support for the old default version from TF 1.x.

[rendered document](https://github.com/kamil-andrzejewski/oneDNN/tree/rfcs/rfcs/20210201-resampling-with-no-pixel-shift)

### CODE REVIEW

**Comment by mgouicem** [comment]
Tagging @kruus @kawakami-k @nSircombe @diaena @alenik01 @agramesh1 @mahmoud-abuzaina @jgong5

File: rfcs/20210201-resampling-with-no-pixel-shift/README.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+not require so many changes in API as it is done in Option 1.
+- Separation of an implementation that is considered inaccurate.
+
+Cons:
+- In case of a request to add support for another algorithm with that behavior,
+we would have to add another algorithm to the list(example: algorithm nearest neighbour without
+half_pixel_centers behavior).
+
+### Recommendation
+I would recommend Option 2. The flag approach seems to be better since it makes
```
Comment (piotrchmiel) [inline rfcs/20210201-resampling-with-no-pixel-shift/README.md:105]: Personally, I would vote for option number 1. In my opinion it is more extendable taking into consideration future development of resampling. 

What if in the future someone will request another feature ? For example align corners from https://www.tensorflow.org/api_docs/python/tf/compat/v1/image/resize_bilinear.

In option 1 - it will be another enum dnnl_align_corners=0x2. What is more it will be easy to combine features. To enable no_pixel_shift with align corners client will set bit 0 and 1. Additionally features will be available at once for all resampling algorithms linear and nearest_neighbour.
In option 2 adding features and combining it with algorithms requires adding a lot of enumerations - algorithm_name_feature_1_feature_2....feature_n.

File: rfcs/20210201-resampling-with-no-pixel-shift/README.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+/// Flags for resampling primitive.
+enum class resampling_flags : unsigned {
+    none = dnnl_resampling_flags_none,
+    half_pixel_centers = dnnl_half_pixel_centers
+};
+```
+
+``` cpp
+...
+desc(prop_kind aprop_kind, algorithm aalgorithm, resampling_flags flags,
```
Comment (igorsafo) [inline rfcs/20210201-resampling-with-no-pixel-shift/README.md:57]: Flags should be `int` to make possible to pass something like `dnnl_half_pixel_centers | dnnl_something | dnnl_someghing_2`, because the result of this operation is integer value and also the result might not be represented as one of flags.

An example:
```cpp
// main.c
typedef enum {
    foo = 0x1,
    bar = 0x2
} flags_t;

int main() {
    flags_t a = foo & bar; // error
    int b = foo & bar;  // works fine

    return 0;
}

```

```bash
$ g++ main.c
main.c: In function ‘int main()’:
main.c:7:23: error: invalid conversion from ‘int’ to ‘flags_t’ [-fpermissive]
     flags_t a = foo & bar;
```

Comment (kamil-andrzejewski) [reply rfcs/20210201-resampling-with-no-pixel-shift/README.md]: I changed to `unsigned`.

File: rfcs/20210201-resampling-with-no-pixel-shift/README.md (hunk 1 above)
Comment (mgouicem) [reply rfcs/20210201-resampling-with-no-pixel-shift/README.md:105]: I agree option 1 is more flexible, but with that flexibility comes more complexity (see the batchnorm precedent).
I am personally in favor of option 2 since it will be simpler to implement/test, and because we are not planning to provide more extensions to resampling (TF2 removed support for this resampling flavor, and other frameworks do not use this 
 definition as it is "broken").

Here, because this oneDNN feature extension is about supporting legacy use cases, I would favor simplicity over flexibility. If it turns out we end up with more flavors over time, we can still consolidate them at that point.

Comment (igorsafo) [reply rfcs/20210201-resampling-with-no-pixel-shift/README.md:105]: While initially I proposed option 1, i would agree with Mourad and vote for option 2. If it affects us in the future and we will see more properties required to be supported we can have resampling_v2.

Right now the second option is straight-forward {API, implementation} -wise and we have a plan for the case in which we need to extend API.

Sorry for the burden.

File: rfcs/20210201-resampling-with-no-pixel-shift/README.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+### Recommendation
+I would recommend Option 2. The flag approach seems to be better since it makes
+the  API more stable, but potentially we can reach a point with combinations we
+don't support. It can confuse users because the combination of some flags and
+algorithms will not work and the user will not know why.
+
+## Open questions
+
+1. I'm not sure if name dnnl_resampling_linear_no_pixel_shift is the best one.
```
Comment (igorsafo) [inline rfcs/20210201-resampling-with-no-pixel-shift/README.md:112]: I looked through [Resampling documentation](https://oneapi-src.github.io/oneDNN/dev_guide_resampling.html) and [naming conventions](https://oneapi-src.github.io/oneDNN/dev_guide_conventions.html) and there is no term "pixel". I would vote for having `dnnl_resampling_linear_no_shift`, but i am open to discussion :)
Casting arch team (@mgouicem @densamoilov @sbogusev @dzarukin )

**Review by igorsafo — APPROVED** [review]

File: rfcs/20210201-resampling-with-no-pixel-shift/README.md (hunk 3 above)
Comment (kamil-andrzejewski) [reply rfcs/20210201-resampling-with-no-pixel-shift/README.md]: Done.

File: rfcs/20210201-resampling-with-no-pixel-shift/README.md (hunk 1 above)
Comment (mgouicem) [reply rfcs/20210201-resampling-with-no-pixel-shift/README.md:105]: Since there is no recent activity in this thread, I will assume that we are ok with option 2.
